graylog2-server/src/main/java/org/graylog/security/shares/EntitySharesService.java

```diff
@@ -139,11 +169,46 @@ public EntityShareResponse prepareShare(EntityShareRequest request,
                 .availableGrantees(modifiableGrantees)
                 .availableCapabilities(getAvailableCapabilities())
                 .activeShares(ImmutableSet.of())
-                .selectedGranteeCapabilities(request.selectedGranteeCapabilities().orElse(ImmutableMap.of()))
+                .selectedGranteeCapabilities(ImmutableMap.of())
                 .validationResult(new ValidationResult())
                 .build();
     }
 
+    /**
+     * Similar to generic prepare without a specific entity; but also checks for missing permissions on specified
+     * dependent entities.
+     * @param entityGRNs check permissions on these entities
+     * @param sharingUser         the sharing user
+     * @return eligible grantees and missing permissions on dependencies
+     */
+    public EntityShareResponse prepareShare(List<String> entityGRNs, User sharingUser) {
+        requireNonNull(entityGRNs, "entityGRNs cannot be null");
+        requireNonNull(sharingUser, "sharingUser cannot be null");
+        final EntityShareResponse response = prepareShare(sharingUser);
+
+        return response.toBuilder()
+                .missingPermissionsOnDependencies(missingPermissions(
+                        response.availableGrantees().stream().map(Grantee::grn).collect(Collectors.toSet()),
+                        entityGRNs, response.sharingUser()))
+                .build();
+    }
+
+    private Map<GRN, Collection<EntityDescriptor>> missingPermissions(Set<GRN> selectedGrantees, List<String> entityGRNs, GRN sharingUserGRN) {
+        final ImmutableSet<EntityDescriptor> entities = entityGRNs.stream()
+                .map(grn -> EntityDescriptor.create(grnRegistry.parse(grn), grn, Set.of()))
+                .collect(ImmutableSet.toImmutableSet());
+
+        final ImmutableMultimap<GRN, EntityDescriptor> deniedEntities =
+                entityDependencyPermissionChecker.check(sharingUserGRN, entities, selectedGrantees);
+
+        Map<GRN, Collection<EntityDescriptor>> missingPermissions = new HashMap<>();
+        deniedEntities.keySet().forEach(grantee -> {
+            final ImmutableSet<EntityDescriptor> missing = deniedEntities.get(grantee).stream().collect(ImmutableSet.toImmutableSet());
+            missingPermissions.put(grantee, missing);
+        });
+        return missingPermissions;
```

Comment:
```suggestion
        return deniedEntities.asMap().entrySet().stream()
                .collect(Collectors.toMap(
                        Map.Entry::getKey,
                        entry -> entry.getValue().stream().collect(ImmutableSet.toImmutableSet())
                ));
```
I would remove the usage of the intermediate Map here